### PR TITLE
Issue 787: Org short names in django admin

### LINKED ISSUE #787: Create shortened names for the Organization Types
### Story

As an Analysts
I want easily readable Organization Types
so that I can quickly sort and filter them.

### Acceptance Criteria

- The options for the Organization Type will read as follows
* Federal
* Interstate
* State or territory
* Tribal
* County
* City
* Special district
* School district

The labels in the table will be the same

### Additional Context

Note for Implementer: Ensure that this is not changed on the actual domain application, 
so when the domain application is going through the registrant does not see the shortened string

### Issue Links

_No response_

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #787 

## Changes

- Use shortened names for orgs in the domain_application and domain_information choicefields which satisfies the admin requirement
- Use another class literal in the application form to display a more verbose value
- Use a filter on the application summary and application manage screens to swap out the short description for the verbose on on those pages
- Unit tests
- Migration
- Bonus: slight edit to the group column in the user table in admin to display a value consistent with the filter and form detail (@Katherine-Osos)

## Context for reviewers

We want to see short organization names (e.g. Federal) in django admin, while keeping the verbose names on user-facing pages (Federal: an agency of the U.S. government's executive, legislative, or judicial branches)

## Code Review Verification Steps

Check admin lists and the admin detail view (form) for Domain Application and Domain Information. The lists display the short name in the table column and filter value. As a side effect of the implementation, the forms use the short names in the dropdown.

EDIT: check the admin list for Domains as well. Those guys pull the org type too, albeit the long way (through the linked domain_info)

Check the application on the user-facing app, check the review page and the application summary page (for a, e.g., submitted application when you click 'manage').

You can check that the value on the review page still says 'Incomplete' when it's incomplete by starting a new application and swapping out register/**review**/ in the url bar.

@Katherine-Osos check the Users tables in admin.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [x] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [x] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [x] Reviewed this code and left comments
- [x] Checked that all code is adequately covered by tests
- [x] Made it clear which comments need to be addressed before this work is merged
- [x] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [x] All new functions and methods are commented using plain language
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Meets all designs and user flows provided by design/product
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [x] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [x] Chrome
  - [ ] Microsoft Edge
  - [x] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

Screen Shot 2023-10-12 at 3 46 13 PM

Screen Shot 2023-10-12 at 3 46 28 PM

Screen Shot 2023-10-12 at 3 46 01 PM

Screen Shot 2023-10-12 at 3 45 50 PM

Screen Shot 2023-10-12 at 3 45 37 PM

### CODE REVIEW

File: src/registrar/templatetags/custom_filters.py [hunk 1]
```diff
@@ -48,3 +49,17 @@ def contains_checkbox(html_list):
         if re.search(r'<input[^>]*type="checkbox"', html_string):
             return True
     return False
+
+
+@register.filter
+def get_organization_long_name(organization_type):
+    organization_choices_dict = {}
+
+    for name, value in DomainApplication.OrganizationChoicesVerbose.choices:
```
Comment (neilmb) [inline src/registrar/templatetags/custom_filters.py:58]: **comment (non-blocking)**: This is idiomatically `dict(DomainApplication.OrganizationChoicesVerbose.choices)`. e.g. https://gist.github.com/OmenApps/3eef60ba4204f3d1842d9d7477efcce1#file-django_choices-txt-L28

File: src/registrar/templatetags/custom_filters.py [hunk 2]
```diff
@@ -48,3 +49,17 @@ def contains_checkbox(html_list):
...
+        organization_choices_dict[name] = value
+
+    long_form_type = organization_choices_dict[organization_type]
+    if long_form_type is not None:
+        return long_form_type
+
+    return "Error"
```
Comment (neilmb) [inline src/registrar/templatetags/custom_filters.py:65]: **suggestion (non-blocking)**: Since this filter is just for organization names, could this be less generic" "Organization type error" maybe?

File: src/registrar/tests/test_admin.py [hunk 3]
```diff
@@ -300,6 +300,23 @@ def setUp(self):
...
+        response = self.client.get("/admin/registrar/domainapplication/")
+        # There are 3 template references to Federal (3) plus one reference in the table
+        # for our actual application
+        self.assertContains(response, "Federal", count=4)
+        # This may be a bit more robust
+        self.assertContains(
+            response, '<td class="field-organization_type">Federal</td>', count=1
+        )
+        # Now let's make sure the long description does not exist
+        self.assertNotContains(response, "Federal: an agency of the U.S. government")
```
Comment (neilmb) [inline src/registrar/tests/test_admin.py:318]: **comment (non-blocking)**: Love this test!

**Review by neilmb — COMMENTED** [review]
I think with the test cases there isn't any reason not to use the "through the window" approach here.

File: src/registrar/tests/test_views.py [hunk 4]
```diff
@@ -1065,6 +1069,26 @@ def test_application_edit_restore(self):
         # page = self.app.get(url)
         # self.assertNotContains(page, "VALUE")
 
+    def test_long_org_name_in_application(self):
+        """
+        Make sure the long name is displaying in the application form,
+        org step
+        """
+        request = self.app.get(reverse("application:")).follow()
+        self.assertContains(request, "Federal: an agency of the U.S. government")
```
Comment (zandercymatics) [inline src/registrar/tests/test_views.py:1078]: (optional) I think for coverage purposes it would be helpful to test on at least one or two additional long org names here.

Comment (rachidatecs) [reply src/registrar/tests/test_views.py:1078]: I feel like if one works then they all certainly work?

File: src/registrar/templatetags/custom_filters.py (hunk 2 above)
Comment (rachidatecs) [reply src/registrar/templatetags/custom_filters.py:65]: You know. I think in context it does not matter (a user sees Organization: Error vs Organization: Organization type error). But maybe we log a organization type error...

File: src/registrar/tests/test_views.py (hunk 4 above)
Comment (zandercymatics) [reply src/registrar/tests/test_views.py:1078]: It'd just be for the very rare edge case of if the dict itself gets messed with but not that value in particular. For instance, someone accidently replaces the entire dict with only the first record. Very much (optional) though, I think its stretching a bit

**Comment by erinysong** [comment]
Not sure if this is linked to my sandbox issues (ignore if other people have gotten this to work), but when I log onto the sandbox, I'm still getting the full names for the org names. Is there a migration that would update the names on the sandbox, or is this working for other people?
**Admin views**
image
image
image
image

**Analyst views** (same long form org name as admin)
image
image

**Comment by rachidatecs** [comment]
Sandbox issue for sure @erinysong, test locally if the code is no longer on my sandbox

File: src/registrar/migrations/0039_merge_20231013_2029.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Generated by Django 4.2.1 on 2023-10-13 20:29
```
Comment (therealslimhsiehdy) [inline src/registrar/migrations/0039_merge_20231013_2029.py:1]: **comment (blocking):** I know, I know migrations are the bane of our existence 🤣 However, I don't think we should be merging migrations files and each migration should be it's own separate migration.

I think this happened when you merged a PR from main into your working branch and bc the migration numbers were the same it merges like this (ie main was at 0038 and you also had 0038 on your local for a migration you made). I believe you can get past the merging of migration numbers clashing by rolling back the migrations back to the most recent migration in main and then rerunning migrations command again so the changes you made are reflected -- feel free to ping me and I can find the commands and do it with you if it's confusing!

**Review by neilmb — COMMENTED** [review]
new changes LGTM

File: src/registrar/forms/application_wizard.py [hunk 6]
```diff
@@ -153,7 +153,8 @@ def from_database(cls, obj: DomainApplication, join: str, on_fetch: Callable):
 
 class OrganizationTypeForm(RegistrarForm):
     organization_type = forms.ChoiceField(
-        choices=DomainApplication.OrganizationChoices.choices,
+        # use the long names in the application form
+        choices=DomainApplication.OrganizationChoicesVerbose.choices,
```
Comment (zandercymatics) [inline src/registrar/forms/application_wizard.py:157]: (Q) do we still need this after the change? Its looks like we probably do but just to make sure its not a leftover

File: src/registrar/templatetags/custom_filters.py [hunk 7]
```diff
@@ -48,3 +51,18 @@ def contains_checkbox(html_list):
         if re.search(r'<input[^>]*type="checkbox"', html_string):
             return True
     return False
+
+
+@register.filter
+def get_organization_long_name(organization_type):
+    # https://gist.github.com/OmenApps/3eef60ba4204f3d1842d9d7477efcce1#file-django_choices-txt-L28
```
Comment (zandercymatics) [inline src/registrar/templatetags/custom_filters.py:58]: (blocking) This link is broken for me... I'd also say that this link needs a bit more context (brief explanation as to what the link is).

If the link is no longer needed you can delete this, also

File: src/registrar/templatetags/custom_filters.py [hunk 8]
```diff
@@ -48,3 +51,18 @@ def contains_checkbox(html_list):
...
+    organization_choices_dict = dict(
+        DomainApplication.OrganizationChoicesVerbose.choices
+    )
+
+    long_form_type = organization_choices_dict[organization_type]
+    if long_form_type is None:
+        logger.error("Organization type error, triggered by a template's custom filter")
```
Comment (zandercymatics) [inline src/registrar/templatetags/custom_filters.py:65]: (optional) We can afford to be more specific here given that this log is for internal uses. If the almighty linter allows, I'd recommend including some metadata about why the error occurred (basically: "long_form_type was None" or "org type cannot be none")

File: src/registrar/tests/test_admin.py [hunk 9]
```diff
@@ -52,6 +52,26 @@ def setUp(self):
...
+
+        response = self.client.get("/admin/registrar/domain/")
+
+        # There are 3 template references to Federal (3) plus one reference in the table
+        # for our actual application
+        self.assertContains(response, "Federal", count=4)
+        # This may be a bit more robust
+        self.assertContains(
+            response, '<td class="field-organization_type">Federal</td>', count=1
+        )
```
Comment (zandercymatics) [inline src/registrar/tests/test_admin.py:71]: Nice, I like the approach here

File: src/registrar/tests/test_admin.py [hunk 10]
```diff
@@ -243,8 +263,11 @@ def test_place_and_remove_hold_epp(self):
         raise
 
     def tearDown(self):
-        User.objects.all().delete()
         super().tearDown()
+        Domain.objects.all().delete()
+        DomainInformation.objects.all().delete()
+        DomainApplication.objects.all().delete()
+        User.objects.all().delete()
```
Comment (zandercymatics) [inline src/registrar/tests/test_admin.py:270]: (Unrelated to your PR / Informational) I don't think it concerns this test but just in general, when your dealing with deleting domains (and by extension DomainInformation) I've noticed that if you have any lingering PublicContact objects, you can't do this directly as you need to delete those contacts first, from a table perspective. 

If this class ever (in one of the test cases), for some reason, wants to test on a domain with PublicContacts this teardown _will_ fail (security.gov / freeman.gov + MockEppLib inherit). That said, we aren't and I do not think this is an issue worth addressing.

**Review by zandercymatics — APPROVED** [review]
Love the approach, looks great. Only thing I noticed is django weirdness. I do not know why, but for domain application when I click "state or territory" (and only that), the screen briefly flashes white like 85% of the time. This does not happen on any other fields nor on domain.

Not going to block because this just seems like django being weird, but I'd be interested to see if you are running into the same thing.

Tested with ANDI and all is good; works with chrome, firefox, and chromium

image

File: src/registrar/forms/application_wizard.py (hunk 6 above)
Comment (rachidatecs) [reply src/registrar/forms/application_wizard.py:157]: yep

File: src/registrar/migrations/0039_merge_20231013_2029.py (hunk 5 above)
Comment (rachidatecs) [reply src/registrar/migrations/0039_merge_20231013_2029.py:1]: A migration merge is one of the relatively easier ways to fix a blown sandbox, but I do get your point. Merges are harder to read. I'll work on it tomorrow.

Comment (rachidatecs) [reply src/registrar/migrations/0039_merge_20231013_2029.py:1]: And I'll ping you. Migrations...

**Review by therealslimhsiehdy — APPROVED** [review]
YAY FOR ONE MIGRATION AND NO MIGRATION MERGES!! Lgtm!!!
